src/main/java/frc/robot/subsystems/swerve/module/Module.java

```diff
@@ -199,37 +101,35 @@ public void setTargetOpenLoopVelocity(double targetVelocityMetersPerSecond) {
 	 * @return the position of the module at the given odometry update index
 	 */
 	public SwerveModulePosition getOdometryPosition(int odometryUpdateIndex) {
-		return new SwerveModulePosition(
-			driveInputs.positionsMeters[odometryUpdateIndex],
-			steerStuff.positionSignal().asArray()[odometryUpdateIndex]
-		);
+		return new SwerveModulePosition(driveInputs.positionsMeters[odometryUpdateIndex], getSteerPositions()[odometryUpdateIndex]);
 	}
 
 	public int getNumberOfOdometrySamples() {
-		return Math.min(driveInputs.positionsMeters.length, steerStuff.positionSignal().asArray().length);
+		return Math.min(getDrivePositions().length, getSteerPositions().length);
 	}
 
 	public SwerveModuleState getTargetState() {
 		return targetState;
 	}
 
 	public SwerveModuleState getCurrentState() {
-		return new SwerveModuleState(getDriveVelocityMetersPerSecond(), getSteerPosition());
+		return new SwerveModuleState(getDriveVelocityMetersPerSecond(), getSteerLatestPosition());
 	}
 
-	public Rotation2d getDrivePosition() {
-		return driveInputs.uncoupledPositions[driveInputs.uncoupledPositions.length - 1];
+	public Rotation2d getSteerLatestPosition() {
+		Rotation2d[] steerPositions = getSteerPositions();
+		return steerPositions[steerPositions.length - 1];
 	}
 
-	public double getDriveVelocityMetersPerSecond() {
-		return driveInputs.velocityMetersPerSecond;
+	public Rotation2d getDriveLatestPosition() {
+		Rotation2d[] drivePositions = getDrivePositions();
+		return drivePositions[drivePositions.length - 1];
 	}
 
-	public Rotation2d getSteerPosition() {
-		return steerStuff.positionSignal().getLatestValue();
+	public double getDriveVelocityMetersPerSecond() {
+		return toDriveMeters(getDriveVelocitySeconds());
```

Comment:
what do u mean?

Reply:
from what I've seen, you're using a different implementation (now with `uncoupleDriveAngle` too) so it may break the odometry. I would test it with a recent version of poseestimator.

Reply:
I don't thinks so, if you're sure the data is exactly the same it's fine, but last time you've done major changes to the swerve that changed the odometry, it took a while to debug it. So I think it's better to test each change to the swerve by itself, to make sure it doesn't happen again.

Reply:
i just added a step in function in the middle instead of two steps in same func. if u see something specific here that maybe i didnt notice let me know